phoenix-core/src/main/java/org/apache/phoenix/schema/types/PJsonDataType.java

```diff
@@ -0,0 +1,258 @@
+
+	PJsonDataType() {
+		super("JSON", Types.OTHER, PhoenixJson.class, null, 48);
+	}
+
+	@Override
+	public int toBytes(Object object, byte[] bytes, int offset) {
+
+		if (object == null) {
+			return 0;
+		}
+		byte[] b = toBytes(object);
+		System.arraycopy(b, 0, bytes, offset, b.length);
+		return b.length;
+
+	}
+
+	@Override
+	public byte[] toBytes(Object object) {
+		if (object == null) {
+			return ByteUtil.EMPTY_BYTE_ARRAY;
+		}
+		PhoenixJson phoenixJson = (PhoenixJson) object;
+		return PVarchar.INSTANCE.toBytes(phoenixJson.toString());
+	}
+
+	@Override
+	public Object toObject(byte[] bytes, int offset, int length,
+			@SuppressWarnings("rawtypes") PDataType actualType,
+			SortOrder sortOrder, Integer maxLength, Integer scale) {
+
+		if (!actualType.isCoercibleTo(this)) {
+			throwConstraintViolationException(actualType, this);
+		}
+		if (length == 0) {
+			return null;
+		}
+		return getPhoenixJson(bytes, offset, length);
+
+	}
+
+	@Override
+	public Object toObject(Object object,
+			@SuppressWarnings("rawtypes") PDataType actualType) {
+		if (object == null) {
+			return null;
+		}
+		if (equalsAny(actualType, PJsonDataType.INSTANCE)) {
+			return object;
+		}
+		if (equalsAny(actualType, PVarchar.INSTANCE)) {
+			return getJsonFromVarchar(object, actualType);
+		}
+		return throwConstraintViolationException(actualType, this);
+	}
+
+	@Override
+	public boolean isCoercibleTo(
+			@SuppressWarnings("rawtypes") PDataType targetType) {
+		return equalsAny(targetType, this, PVarchar.INSTANCE);
+
+	}
+
+	@Override
+	public boolean isCoercibleTo(
+			@SuppressWarnings("rawtypes") PDataType targetType, Object value) {
+		return isCoercibleTo(targetType);
+	}
+
+	@Override
+	public boolean isSizeCompatible(ImmutableBytesWritable ptr, Object value,
+			@SuppressWarnings("rawtypes") PDataType srcType, Integer maxLength,
+			Integer scale, Integer desiredMaxLength, Integer desiredScale) {
+		if (ptr.getLength() != 0 && maxLength != null
+				&& desiredMaxLength != null) {
+			return maxLength <= desiredMaxLength;
+		}
+		return true;
+	}
+
+	@Override
+	public boolean isFixedWidth() {
+		return false;
+	}
+
+	@Override
+	public int estimateByteSize(Object o) {
+		PhoenixJson phoenixJson = (PhoenixJson) o;
+		return phoenixJson.toString().length();
+	}
+
+	@Override
+	public Integer getByteSize() {
+		return null;
+	}
+
+	@Override
+	public int compareTo(Object lhs, Object rhs,
+			@SuppressWarnings("rawtypes") PDataType rhsType) {
+		if (PJsonDataType.INSTANCE != rhsType) {
+			throwConstraintViolationException(rhsType, this);
+		}
+		PhoenixJson phoenixJsonLHS = (PhoenixJson) lhs;
+		PhoenixJson phoenixJsonRHS = (PhoenixJson) rhs;
+		return PVarchar.INSTANCE.compareTo(phoenixJsonLHS.toString(),
+				phoenixJsonRHS.toString(), rhsType);
+	}
+
+	@Override
+	public Object toObject(String value) {
+		byte[] jsonData = Bytes.toBytes(value);
+		return getPhoenixJson(jsonData, 0, jsonData.length);
```

Comment:
You don't want to go from String->byte[]->String here. Create a getPhoenixJson(String) method.